### PR TITLE
Feature: Custom Node Colors for Models

### LINKED ISSUE #44: [CT-960] Configurable Color-Coding of Schemas/Nodes in Docs
Similar to [here](https://github.com/fishtown-analytics/dbt-docs/issues/26).
Essentially, we are interested in being able to color code either schemas or individual models in the dbt docs in a similar fashion to the way you can configure materialization of individual models.

e.g. 
```
models:
  rad_warehouse:
    materialized: view
    staging:
      schema: staging
      materialized: view
      color: #453453
    analytics:
      schema: analytics
      materialized: table
      color: #589593
```
We really like the difference in coloration between source and models, and it would be awesome to carry this further to make it easier to distinguish between different parts of the dag.

A colleague (Taylor M.) and I are interested in working on this. We can imagine implementing this in a couple of different ways, either through the dbt_project.yml, or through configurable tags in models or schemas.

Do you have any suggestions, tips, warnings, or words of caution?

Thanks!

### PR DESCRIPTION
**Pull request by sungchun12**
resolves #44

### Description

When looking at a DAG, there are a couple of visually oriented questions that immediately come to mind:
- Can I understand the flow at a higher level than the individual nodes?
- E.g. “layers”, or bronze/silver/gold etc.
- The above shows a trend in which dbt is being asked to manage more types of objects than tables and views at a surface level. That being the case, it makes sense to make it easier to reason about those various DAG components more flexibly than we do today.

I'm able to get the node custom color dynamically mapped to the right nodes and with dynamic colors to boot! Similar to what's accomplished in #275 

 - [ ] This code will only work once this PR is merged: https://github.com/dbt-labs/dbt-core/pull/5343

Notion link for my scratchpad troubleshooting: [here](https://www.notion.so/dbtlabs/Update-Custom-Coloring-Logic-8ddfff8355fa400daa367d7370e79ad7)

image

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have generated docs locally, and this change appears to resolve the stated issue
 - [ ] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: src/app/services/graph.service.js [hunk 1]
```diff
@@ -78,6 +78,7 @@ angular
             },
             elements: [],
             layout: layouts.none,
+            node_color: '#000000',
```
Comment (sungchun12) [inline src/app/services/graph.service.js:81]: was thinking through setting up a `node_color` var to then override later within the `setNodes` function

File: src/app/services/graph.service.js [hunk 2]
```diff
@@ -299,15 +312,33 @@ angular
...
+
+            // if (el.data.docs && el.data.docs.node_color) {
+            //     el.data['node_color'] = el.data.docs.node_color;
+            // }
+            
         });
         service.graph.elements = _.filter(elements, function(e) { return e.data.display == 'element'});
-
+        // console.log('hello node_ids  ' + node_ids);
+        console.log('hello elements ' + JSON.stringify(_.each(elements, function(e) { return e.data.node_color}))); //get the node_color isolated in the console logs
```
Comment (sungchun12) [inline src/app/services/graph.service.js:339]: This shows the full json object of node configs which includes the `node_color`

File: src/app/services/graph.service.js [hunk 3]
```diff
@@ -191,6 +192,15 @@ angular
                         'background-color': '#ff5688',
                     }
                 },
+                // apply custom colors on top of all defaults, but allow the selected and hidden options below to override this
+                // TODO: How to get the custom color from the dictionary mapped below
+                {
+                    selector: 'node[node_color_indicator=1]', //this is correctly mapped to the node_color attribute
+                    style: {
+                        'background-color': '#000000', //placeholder hex color, this needs to be dynamic
```
Comment (sungchun12) [inline src/app/services/graph.service.js:200]:
```suggestion
                        'background-color': node_color_variable, //placeholder hex color, this needs to be dynamic
```

Comment (sungchun12) [reply src/app/services/graph.service.js:200]: Even if this works to create a dynamic color mapping, it can only do it with ONE color at a time, not many concurrently. 

I suspect the selector for nodes with the custom color config will require creating a dynamic mapping of unique nodes to unique colors. Pseudo-code below.
```javascript
                {
                    selector: 'node[name="dim_parts"]',
                    style: {
                        'background-color': node_color_var[name="dim_parts"], // dynamic parameter
                    }
                },

```

**Comment by sungchun12** [comment]
Relevant slack thread: [here](https://dbt-labs.slack.com/archives/C0131TY7EEA/p1654716999868699)

Emily is a good point of contact for Javascript help

**Comment by drewbanin** [comment]
@sungchun12 can you give me an idea of what the model configuration you have in mind looks like? Is it:
```
models:
  - name: dim_customers
    docs:
      show: true
      color: #c0ffee
```

Happy to help with the JS stuff in here (to the extent that I'm able) but also wondering if we want to consider coloring nodes in some sort of grouping (eg. by tag? or something else?). Just thoughts!

**Comment by sungchun12** [comment]
https://github.com/dbt-labs/dbt-docs/pull/280#discussion_r893882591

File: src/app/services/graph.service.js [hunk 4]
```diff
@@ -299,15 +308,22 @@ angular
 
             if (highlight && _.includes(highlight, el.data.unique_id)) {
                 el.data['selected'] = 1;
+                console.log('hello selected node color  ' + el.data['selected']);
             }
 
             if (el.data.docs && el.data.docs.show === false) {
-                el.data['hidden'] = 1;
+                el.data['hidden'] = 1; //set to 1 to set as hidden
```
Comment (sungchun12) [inline src/app/services/graph.service.js:315]:
```suggestion
                el.data['hidden'] = 1;
```

File: src/app/services/graph.service.js [hunk 5]
```diff
@@ -299,15 +308,22 @@ angular
...
+            if (el.data.docs && el.data.docs.node_color) {
+                el.data['node_color'] = el.data.docs.node_color;
+                console.log('hello docs node_color  ' + el.data['node_color']);
+            }
+            
         });
         service.graph.elements = _.filter(elements, function(e) { return e.data.display == 'element'});
-
         return node_ids;
+
```
Comment (sungchun12) [inline src/app/services/graph.service.js:326]:
```suggestion
```

File: src/app/services/graph.service.js [hunk 6]
```diff
@@ -191,6 +191,14 @@ angular
                         'background-color': '#ff5688',
                     }
                 },
+                // apply custom colors on top of all defaults, but allow the selected and hidden options below to override this
```
Comment (sungchun12) [inline src/app/services/graph.service.js:194]:
```suggestion
```

File: src/app/services/graph.service.js [hunk 7]
```diff
@@ -216,6 +223,7 @@ angular
                         'background-opacity': 0.5,
                     }
                 },
+
```
Comment (sungchun12) [inline src/app/services/graph.service.js:226]:
```suggestion
```

File: src/app/services/graph.service.js [hunk 8]
```diff
@@ -191,6 +191,13 @@ angular
                         'background-color': '#ff5688',
                     }
                 },
+                {
+                    selector: 'node[node_color]', 
+                    style: {
+                        'background-color': 'data(node_color)', 
+                    }
+
```
Comment (sungchun12) [inline src/app/services/graph.service.js:199]:
```suggestion
```

**Comment by sungchun12** [comment]
I got this to work!
image

**Comment by sungchun12** [comment]
@b-per will own testing this feature!

**Comment by sungchun12** [comment]
@b-per can you update the changelog per the below?

> I have updated the CHANGELOG.md and added information about my change to the "dbt next" section.

**Review by gwenwindflower — APPROVED** [review]
:shipit:

**Comment by b-per** [comment]
I have tested this code in conjunction with https://github.com/dbt-labs/dbt-core/pull/5343

**tldr;**
The functionality works, but in the case of invalid colors, the DAG doesn't render properly. I don't know if input verification should be handled on the dbt Core side, the dbt docs side, or both.

## ✅ Defining valid color names and valid hex codes is working correctly
Core Side
image
Docs Side
image

## ✅ Defining the same color twice works correctly
Core Side
image
Docs Side
image

## ✅ Defining a node hidden and giving it a color still keeps it grey, as intended
Core Side
image
Docs Side
image

## ⛔ Issues

### dbt docs behavior when there are issues with the `node_color` input
The issues I saw are related to incorrect data input (example with an hex code without the starting `#`)

Error in Dev Tools:
image
How the DAG appears:
image
How the DAG appears after reloading the page (empty)
image

### variations of node_color tested
- ⛔ Fails
  - an invalid color name: `emerald`
  - an hexcode without the hash: `111111`
  - a string made of only spaces (but not empty): `"  "`
  - typing the value `null`
- ✅ Works
  - A valid color name
  - A valid hex code
  - A valid color name or Hex code surrounded with spaces: `" #1111   "`
  - An empty string (keeps the default color): `""`
  - Defining a custom color on models and seeds

**Review by stu-k — COMMENTED** [review]
Defining an invalid `background-color` (as @b-per has listed above) will not set the `background-color` property on the element. This produces the error `The style property 'background-color: whatever' is invalid`. Because this isn't a valid style property, it is not set on the element, leading into the `node_color` issue.

It's been a while since I've been in angular, but I believe the lines
```
                {
                    selector: 'node[node_color]', 
                    style: {
                        'background-color': 'data(node_color)', 
                    }
                },
```
are causing this, as an invalid `background-color` will not be set.

I think this selector should be moved from this magical angular list to the place below where we iterate over each of the graph elements.

This gives us the ability to validate the color. [Something like this should be sufficient](https://gist.github.com/stu-k/aac9d0fb7bf5e325f4fa8d260bd28fc8).
